http-response/background.js

```diff
@@ -1,22 +1,154 @@
 function listener(details) {
+  // The received data is a stream of bytes. In order to do text-based
+  // modifications, it is necessary to decode the bytes into a string
+  // using the proper character encoding, do any modifications, then
+  // encode back into a stream of bytes.
+  // Historically, detecting character encoding has been a tricky task
+  // taken on by the browser. Here, a simplified approach is taken
+  // and the complexity is hidden in a helper method.
+  let decoder, encoder;
+  [decoder, encoder] = detectCharsetAndSetupDecoderEncoder(details);

```

Comment:
The logic is complicated but still not a full solution to the problem.

For simplicity, I recommend to check whether the response is (X)HTML (to ignore responses such as images, media, downloads), and only proceed if needed. Then the MIME/charset detection logic can greatly be simplified (see also comment below about charset detection).

You should also check that the HTTP status code is OK before attaching a stream filter.

Reply:
The note about the complex logic is discussed in the other relevant comments below, but duly noted.

I agree it would be better to move the logic for detecting the supported MIME types as a pre-check here.
- [ ] Pre-check MIME types

Also agreed on the HTTP status code check.

- [ ] Add HTTP status code check